### PR TITLE
C3 nohahesham feature f1

### PR DESCRIPTION
**Pull request by nohahesham**
@m0hamed 
@SalmaFG 
@Omar-Hossam 
issue #95

### CODE REVIEW

File: arability/app/models/my_subscription.rb [hunk 1]
```diff
@@ -3,11 +3,6 @@ class MySubscription < ActiveRecord::Base
   attr_accessible :developer, :word_add, :word_follow, :word_search, :subscription_model_id
   belongs_to :developer
   validates :subscription_model_id, :presence => true
-
-  @@search=1
-  @@add=2
-  @@follow=3
-
   # Author:
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb:7]: please insert an empty line to differentiate between validation and new function

File: arability/app/models/my_subscription.rb [hunk 2]
```diff
@@ -52,105 +47,92 @@ class << self
     #   or follow limit
     # fail:
     #   none
-    
-      def get_permissions(dev_id,type)
+      def get_permission_follow(dev_id)
         my_subscription = 
          MySubscription.joins(:developer).where(:developer_id => dev_id).first
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: follow code conventions, avoid using =>

File: arability/app/models/my_subscription.rb [hunk 3]
```diff
@@ -52,105 +47,92 @@ class << self
...
+  # Author:
+  #  Noha Hesham
+  # Description:
+  #  counts the number of the followed word by the
+  #  developer till now
+  # Success:
+  #  gets the correct number of words counted
+  # Failure:
+  #  none 
    def count_follow
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: please adjust indentation

File: arability/app/models/my_subscription.rb [hunk 4]
```diff
@@ -52,105 +47,92 @@ class << self
     #   or follow limit
     # fail:
     #   none
-    
-      def get_permissions(dev_id,type)
+      def get_permission_follow(dev_id)
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: allign line with rest of code, and re-allign rest of function

File: arability/app/models/my_subscription.rb [hunk 5]
```diff
@@ -52,105 +47,92 @@ class << self
...
+  #   limit and returns true if the developer didnt pass the limit
+  # Success:
+  #   Gives permission
+  # Failure:
+  #   None 
+  def self.get_projects
+    dev = Developer.find_by_gamer_id(current_gamer.id).id
+    my_subscription = 
+      MySubscription.joins(:developer).where(:developer_id => dev_id).first
+      project = Project.where(:developer_id => owner_id)
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: follow code conventions and avoid using => for the previous 2 lines

Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: allign correctly this line and all the next ones with the code

File: arability/app/models/my_subscription.rb [hunk 6]
```diff
@@ -52,105 +47,92 @@ class << self
...
+  #   It compares the words added by the developer to the add
+  #   limit and returns true if the developer didnt pass the limit
+  # Success:
+  #   Gives permission to add words
+  # Failure:
+  #   None 
+  def self.add_max_word
+    dev = Developer.find_by_gamer_id(current_gamer.id).id
+    my_subscription = 
+      MySubscription.joins(:developer).where(:developer_id => dev_id).first
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: avoid using => for this line and the next 2, and allign properly the next lines withe rest of the code

File: arability/app/models/my_subscription.rb [hunk 7]
```diff
@@ -52,105 +47,92 @@ class << self
...
+      project = Project.where(:developer_id => owner_id)
+      add=PreferedSynonym.where(:project_id => project_id )
+      if add.count < my_subscription.limit
+        return true
+      else
+        return false
+  end
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb:131]: I could see only one end which is supposly the end of the function, what about the end of the if?

File: arability/app/models/my_subscription.rb [hunk 8]
```diff
@@ -52,105 +47,92 @@ class << self
...
+  #   gives permission accordingly
+  # Success:
+  #   Gives permission to search
+  # Failure:
+  #   None 
+  def self.get_max_words(word_id)
+    dev_id = Developer.find_by_gamer_id(current_gamer.id).id
+    my_subscription = 
+      MySubscription.joins(:developer).where(:developer_id => dev_id).first
+    word = Search.joins(:keyword).where(:keyword_id => word_id)
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: avoid using => for previous 2 lines

File: arability/app/models/my_subscription.rb [hunk 9]
```diff
@@ -52,105 +47,92 @@ class << self
...
+    if word!= nil
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb:148]: leave space between operands and operators

File: arability/app/models/my_subscription.rb [hunk 10]
```diff
@@ -52,105 +47,92 @@ class << self
...
+      return true
+    else 
+      if my_subscription.word_search > Search.where(:developer_id => dev_id).count 
+         search=Search.new
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: line is indented 3 spaces more than the previous, while it's supposed to be indented by 2 extra spaces only. Please allign this line and the next lines properly with code

File: arability/config/routes.rb [hunk 11]
```diff
@@ -53,12 +53,11 @@
 
     scope "developers/" do 
       match "/" => "backend#home", :as => "backend_home"
-
+      match 'projects/remove_developer_from_project' => 'developer#remove_developer_from_project'
```
Comment (omar-mohyeldin) [inline arability/config/routes.rb:93]: use double quotes

File: arability/spec/models/my_subscription_spec.rb [hunk 12]
```diff
@@ -32,17 +31,9 @@
...
-		result = my_sub.decrement_word_add
-		expect(result).to eq(true)
-	end
-
-  it "Word search can de decremented" do
-		result = my_sub.decrement_word_search
-		expect(result).to eq(true)
-	end
-
+  end
```
Comment (omar-mohyeldin) [inline arability/spec/models/my_subscription_spec.rb:57]: leave an empty line between each 2 tests

File: arability/spec/models/my_subscription_spec.rb [hunk 13]
```diff
@@ -32,17 +31,9 @@
...
+  it "follow permission is given" do 
+    result = MySubscription.get_permission_follow(developer.id)
+    expect(result).to eq(true)
+  end
 end
```
Comment (omar-mohyeldin) [inline arability/spec/models/my_subscription_spec.rb:70]: leave an empty line after each test

**Comment by omar-mohyeldin** [comment]
@nohahesham modify your code

File: arability/app/controllers/my_subscription_controller.rb [hunk 14]
```diff
@@ -1,3 +1,4 @@
+#encoding: UTF-8
 class MySubscriptionController < ApplicationController
   before_filter :authenticate_gamer!
   #before_filter :prepare_subscriptions, :only => [:choose_sub, :pick]
```
Comment (omar-mohyeldin) [inline arability/app/controllers/my_subscription_controller.rb]: remove commented line if this was your code, if not, ask the one who wrote it to remove it in his/her branch

File: arability/app/models/my_subscription.rb [hunk 15]
```diff
@@ -3,11 +3,6 @@ class MySubscription < ActiveRecord::Base
   attr_accessible :developer, :word_add, :word_follow, :word_search, :subscription_model_id
   belongs_to :developer
   validates :subscription_model_id, :presence => true
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: avoid using => if this was your code, if not, ask the one who wrote it to remove it from his/her branch

File: arability/app/models/my_subscription.rb (hunk 1 above)
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb:7]: and check your rest of code for the presence of an empty line between each 2 funtions

File: arability/app/models/my_subscription.rb [hunk 16]
```diff
@@ -52,105 +45,85 @@ class << self
...
-        elsif  type = @@add
-          if my_subscription.word_add == 0
-            return false
-          else
-            return true
-          end 
-        else type = @@follow
-          if @count < my_subscription.word_follow 
+      def get_permission_follow(dev_id)
+          developer = self.developer
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: function code should only be indented with 2 spaces more that the def line of function. Please adjust that here and for the rest of the code

File: arability/app/models/my_subscription.rb [hunk 17]
```diff
@@ -52,105 +45,85 @@ class << self
...
-          end
-        elsif  type = @@add
-          if my_subscription.word_add == 0
-            return false
-          else
-            return true
-          end 
-        else type = @@follow
-          if @count < my_subscription.word_follow 
+      def get_permission_follow(dev_id)
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: why is this code indented with 6 spaces? check your indentation for the whole code

File: arability/app/models/my_subscription.rb [hunk 18]
```diff
@@ -52,105 +45,85 @@ class << self
...
+  # Description:
+  #   It compares the created projects with the developer's project
+  #   limit and returns true if the developer didnt pass the limit
+  # Success:
+  #   Gives permission
+  # Failure:
+  #   None 
+  def get_projects_limit
+    developer = self.developer
+    projects_count = Project.where(:owner_id => developer.id).count
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: avoid using =>

File: arability/app/models/my_subscription.rb [hunk 19]
```diff
@@ -52,105 +45,85 @@ class << self
...
+  # Failure:
+  #  none 
    def count_follow
     @developer = Developer.find(self.developer_id)
-    @count_follow=@developer.Keywords.count
-   end
-end
+    @count_follow=@developer.Keywords.count   
+  end
+  # Author:
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb:73]: leave an empty line between functions

File: arability/app/models/my_subscription.rb [hunk 20]
```diff
@@ -52,105 +45,85 @@ class << self
...
+  #   It compares the words added by the developer to the add
+  #   limit and returns true if the developer didnt pass the limit
+  # Success:
+  #   Gives permission to add words
+  # Failure:
+  #   None 
+  def max_add_word
+    developer = self.developer
+    project = Project.where(:developer_id => owner_id)
+    add=PreferedSynonym.where(:project_id => project_id )
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: avoid using => for the whole code

File: arability/app/models/my_subscription.rb [hunk 21]
```diff
@@ -52,105 +45,85 @@ class << self
...
+  #   if no it checks if the developer has passed the search limit and 
+  #   gives permission accordingly
+  # Success:
+  #   Gives permission to search
+  # Failure:
+  #   None 
+  def self.get_max_words(word_id)
+    developer = self.developer
+    word = Search.joins(:keyword).where(:keyword_id => word_id)
+    if word!= nil
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb:148]: leave spaces between operands and operators

**Comment by omar-mohyeldin** [comment]
@nohahesham please modify your code, and for the comments I've stated, please pass by the whole code to modify repeated mistakes

File: arability/app/models/my_subscription.rb [hunk 22]
```diff
@@ -19,138 +14,135 @@ class MySubscription < ActiveRecord::Base
...
+  # Failure:
+  #   None
+  def get_max_words(word_id)
+    developer = self.developer
+    word = Search.joins(:keyword).where(keyword_id: word_id)
+    if word!= nil
+      return true
+    else
+      if self.word_search > Search.where(developer_id: self.developer).count
+        search=Search.new
```
Comment (m0hamed) [inline arability/app/models/my_subscription.rb]: you should create a new search with the developer id and keyword id and save it in the database

File: arability/app/models/my_subscription.rb [hunk 23]
```diff
@@ -19,138 +14,135 @@ class MySubscription < ActiveRecord::Base
...
+  # Description:
+  #   Returns true if the followed words are less
+  #   than the follow limit,false otherwise
+  # Params:
+  #   None
+  # Success:
+  #   Permission is given if the developer didnt exceed the follow limit
+  # Failure:
+  #   None
+  def get_permission_follow
```
Comment (m0hamed) [inline arability/app/models/my_subscription.rb:47]: this method should count the number of words currently followed and see if they are less than the limit. to count the number of words a developer has followed you can use developer.keywords.count

Comment (nohahesham) [inline arability/app/models/my_subscription.rb:47]: it is counted in another method called count

File: arability/app/models/my_subscription.rb [hunk 24]
```diff
@@ -19,138 +14,140 @@ class MySubscription < ActiveRecord::Base
...
+    developer = self.developer
+    count_follow=developer.Keywords.count
```
Comment (m0hamed) [inline arability/app/models/my_subscription.rb]: developer.keywords.count

the k is small

File: arability/app/models/my_subscription.rb [hunk 25]
```diff
@@ -19,138 +14,140 @@ class MySubscription < ActiveRecord::Base
...
+  #  developer till now
+  # Params:
+  #   None
+  # Success:
+  #  Gets the correct number of words counted
+  # Failure:
+  #  None 
+  def count_follow
+    developer = self.developer
+    count_follow=developer.Keywords.count
```
Comment (m0hamed) [inline arability/app/models/my_subscription.rb]: same here

File: arability/app/models/my_subscription.rb [hunk 26]
```diff
@@ -19,138 +14,140 @@ class MySubscription < ActiveRecord::Base
...
+  #   Noha Hesham
+  # Description:
+  #   It takes the word id and checks if the developer has searched for it before
+  #   if no it checks if the developer has passed the search limit and
+  #   gives permission accordingly
+  # Success:
+  #   Gives permission to search
+  # Failure:
+  #   None
+  def get_max_words(word_id)
```
Comment (m0hamed) [inline arability/app/models/my_subscription.rb]: the name should not be get max words but should be can_search_word

File: arability/app/models/my_subscription.rb [hunk 27]
```diff
@@ -19,138 +14,140 @@ class MySubscription < ActiveRecord::Base
...
+  # Description:
+  #   It compares the words added by the developer to the add
+  #   limit and returns true if the developer didnt pass the limit
+  # Params:
+  #   Proj_id is the id of the project
+  # Success:
+  #   Gives permission to add words
+  # Failure:
+  #   None 
+  def max_add_word(proj_id)
```
Comment (m0hamed) [inline arability/app/models/my_subscription.rb]: this method should return the number of words remaining he could add not true or false

Comment (nohahesham) [inline arability/app/models/my_subscription.rb]: there are 2 one that returns true and false and one that returns the number

File: arability/app/models/my_subscription.rb [hunk 28]
```diff
@@ -19,138 +14,140 @@ class MySubscription < ActiveRecord::Base
...
+  add=PreferedSynonym.where(project_id: proj_id ).count
+  count_num=self.word_add-add
+  return count_num
+  end
+  # Author:
+  #   Noha Hesham
+  # Description:
+  #   It takes the word id and checks if the developer has searched for it before
+  #   if no it checks if the developer has passed the search limit and
+  #   gives permission accordingly
```
Comment (salmafg) [inline arability/app/models/my_subscription.rb:138]: Params?

File: arability/app/models/my_subscription.rb [hunk 29]
```diff
@@ -19,138 +14,140 @@ class MySubscription < ActiveRecord::Base
...
+  # Params:
+  #   None
+  # Success:
+  #  Gets the correct number of words counted
+  # Failure:
+  #  None 
+  def count_follow
+    developer = self.developer
+    count_follow=developer.keywords.count   
+  end
```
Comment (salmafg) [inline arability/app/models/my_subscription.rb:71]: Pleas leave an empty line between each and every method to make it easier to read.

**Comment by salmafg** [comment]
Testing verified.

File: arability/app/models/my_subscription.rb (hunk 22 above)
Comment (NourhanMohamed) [inline arability/app/models/my_subscription.rb]: Shouldn't all the checking be in a model? I see they are all here in the my_subscription controller which are actions that render nothing and redirect no where but return true or false

File: arability/app/models/my_subscription.rb [hunk 30]
```diff
@@ -116,23 +123,26 @@ def max_add_word(proj_id)
   #   Returns number of words 
   # Failure:
   #   None 
-  def max_add_word_count(proj_id)
+  def can_add_word_count(proj_id)
   developer = self.developer
   add=PreferedSynonym.where(project_id: proj_id ).count
   count_num=self.word_add-add
   return count_num
   end
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb:131]: adjust indentation of previous lines to be indented by 2 more spaces than the def line

File: arability/app/models/my_subscription.rb [hunk 31]
```diff
@@ -140,10 +150,12 @@ def get_max_words(word_id)
     else
       if self.word_search > Search.where(developer_id: self.developer).count
         search=Search.new
+        search.developer_id=developer
+        search.synonym_id=word_id
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: please leave spaces between operands and operators for the previous 3 lines

File: arability/app/models/my_subscription.rb [hunk 32]
```diff
@@ -140,10 +150,12 @@ def get_max_words(word_id)
...
+        search.save
         return true
       else
         return false
       end
     end
-  end
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb:153]: this end was for the function and now there is no end for the function, please re add it

File: arability/spec/models/my_subscription_spec.rb [hunk 33]
```diff
@@ -54,12 +55,19 @@
...
     keyword = Keyword.create(name: "click", approved: true)
-    result = my_sub.get_max_words(keyword.id)
+    result = my_sub.can_search_word(keyword.id)
     expect(result).to eq(true)
   end
+
+	it "developer can choose subscription model" do
+		result = MySubscription.choose(developer.id, submodel.id)
+		expect(result).to eq(true)
+
```
Comment (omar-mohyeldin) [inline arability/spec/models/my_subscription_spec.rb]: this line is of no need, please remove it

File: arability/spec/models/my_subscription_spec.rb [hunk 34]
```diff
@@ -54,12 +55,19 @@
...
     expect(result).to eq(true)
   end
    it "search permission is given" do 
     keyword = Keyword.create(name: "click", approved: true)
-    result = my_sub.get_max_words(keyword.id)
+    result = my_sub.can_search_word(keyword.id)
     expect(result).to eq(true)
   end
+
+	it "developer can choose subscription model" do
```
Comment (omar-mohyeldin) [inline arability/spec/models/my_subscription_spec.rb]: this funtion should be indented by 2 more spaces as it is in a "do" loop

File: arability/spec/models/my_subscription_spec.rb [hunk 35]
```diff
@@ -54,12 +55,19 @@
...
+		result = MySubscription.choose(developer.id, submodel.id)
+		expect(result).to eq(true)
+
+	end
+
```
Comment (omar-mohyeldin) [inline arability/spec/models/my_subscription_spec.rb]: there is a missing end for the "do"

**Comment by omar-mohyeldin** [comment]
@nohahesham please modify your code

**Comment by salmafg** [comment]
Documentation verified.

**Comment by abulazm22** [comment]
general notes:
developer = developer.self should be a self refrenced arrow first self.developer then developer
if and else should be inside brackets [ ]
for follow:

for search:
modify Search.where(keyword_id: word_id).first
alt inside boxes
Search.where(developer_id: self.developer.id).count should be calculated first and returned as temp then use it in if as temp
Search.new and search should be on activation box. so is the last returns

add: 
missing title
and call to method

**Comment by abulazm22** [comment]
get_projects_limit:
return true and false should be on activation box

File: arability/app/models/my_subscription.rb [hunk 36]
```diff
@@ -44,12 +45,15 @@ def self.choose(dev_id,sub_id)
   # Failure:
   #   None
   def get_permission_follow
-    if(self.word_follow > 0)
+    developer = self.developer
+    count_follow=developer.keywords.count
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: leave space before and after "="

File: arability/app/models/my_subscription.rb [hunk 37]
```diff
@@ -62,9 +66,10 @@ def get_permission_follow
   # Failure:
   #  None 
   def count_follow
-    @developer = Developer.find(self.developer_id)
-    @count_follow=@developer.Keywords.count   
+    developer = self.developer
+    count_follow=developer.keywords.count
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: leave space before and after "="

File: arability/app/models/my_subscription.rb [hunk 38]
```diff
@@ -96,15 +102,16 @@ def get_projects_limit
   #   Gives permission to add words
   # Failure:
   #   None 
-  def max_add_word(proj_id)
+  def can_add_word(proj_id)
     developer = self.developer
-    add=PreferedSynonym.where(project_id: proj_id )
+    add=PreferedSynonym.where(project_id: proj_id)
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: leave space before and after "="

File: arability/app/models/my_subscription.rb [hunk 39]
```diff
@@ -116,30 +123,36 @@ def max_add_word(proj_id)
...
   # Failure:
   #   None 
-  def max_add_word_count(proj_id)
-  developer = self.developer
-  add=PreferedSynonym.where(project_id: proj_id ).count
-  count_num=self.word_add-add
-  return count_num
+  def can_add_word_count(proj_id)
+    developer = self.developer
+    add=PreferedSynonym.where(project_id: proj_id ).count
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: leave space before and after "="

File: arability/app/models/my_subscription.rb [hunk 40]
```diff
@@ -116,30 +123,36 @@ def max_add_word(proj_id)
...
+    count_num=self.word_add-add
```
Comment (omar-mohyeldin) [inline arability/app/models/my_subscription.rb]: leave space before and after "="

File: arability/spec/models/my_subscription_spec.rb [hunk 41]
```diff
@@ -3,7 +3,8 @@
 
 describe MySubscription do
   let(:developer){
-		developer=Developer.new
+    developer=Developer.new
```
Comment (omar-mohyeldin) [inline arability/spec/models/my_subscription_spec.rb]: leave space before and after "="

**Comment by omar-mohyeldin** [comment]
@nohahesham please modify your code, and pull master to make the pull request mergable to be able to verify

File: arability/spec/models/my_subscription_spec.rb [hunk 42]
```diff
@@ -3,10 +3,14 @@
 
 describe MySubscription do
   let(:developer){
+<<<<<<< HEAD
+    developer = Developer.new
+=======
     developer=Developer.new
+>>>>>>> f73a826ca8817db020623e1429b95be30fe62463
```
Comment (omar-mohyeldin) [inline arability/spec/models/my_subscription_spec.rb]: please solve conflicts

**Comment by omar-mohyeldin** [comment]
@nohahesham almost verify, but I can't fully verify with merge conflicts in your code

**Comment by omar-mohyeldin** [comment]
@nohahesham code verified

**Pull request merged**